### DISCUSSION TITLE
[RFC] Warning filter

### DISCUSSION THREAD

**Discussion by zombieJ** [ant-design/ant-design#44551] [RFCs - archive]
# 📋 RFC

## Summary

ConfigProvider provides `warning` prop to disabled part of warning.

## Basic example

```tsx
<ConfigProvider warning={{ deprecated: false }}>
  <Select><Select.Option /></Select>
  <Form><Form.Item name="name"><Input /><Input /></Form.Item></Form>
</ConfigProvider>
```

Should only warning for Form.Item should not have multiple children with `name` but not warning Select should use `options`.

## Motivation

For development env, some developer upgrade from v4 which may has lots of warning message. It's not easy to find out real error message for breaking change one but mixed with deprecated message.

We can provider the props to enable ConfigProvider disable part of warning message.

## Detailed

Inject `_util/warning` to enable warning message level.

## Drawbacks

Maybe affect user upgrade to next major version.

## Adoption strategy

Apply in minor version.

**Comment by crazyair** [comment]
使用静态方法配置？

**Comment by MadCcc** [comment]
为啥不直接把 deprecated warning 都从 `console.error` 改成 `console.warning`

**Reply by zombieJ** [reply]
以前有讨论过。有几点：
* 和 React warning 保持一致，使用 `console.error`
* deprecated 信息设计上不是可以忽略的，只是因为我们做了兜底所以看似不影响
* 换个地方，并不解决扰屏问题

**Reply by afc163** [reply]
https://github.com/BerkeleyTrue/warning/issues/3
https://github.com/facebook/react/issues/9395

**Comment by MadCcc** [comment]
```ts
interface ConfigProviderProps {
  warning: { priority: 'high' | 'low' }
}
```
`high` meanings every warning matters. `low` means only neccesary warnings will be displayed.
It will simplify this usage, rather than configuring each rules like eslint:

```tsx
<ConfigProvider warning={{ deprecated: false, usage: false, breaking: false }}>
```

**Reply by zombieJ** [reply]
```tsx
interface ConfigProviderProps {
  warning: {[info]?: boolean }
}
```

**Reply by zombieJ** [reply]
名字要再待定一下

**Reply by afc163** [reply]
breaking 最好别给 false 吧。

**Reply by zombieJ** [reply]
不会有 breaking，只会把不必须的关闭。其他的都是必定展示。
